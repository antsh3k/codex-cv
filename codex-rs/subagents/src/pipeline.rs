use serde::Deserialize;
use serde::Serialize;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RequirementsSpec {
    pub title: String,
    pub overview: String,
    pub requirements: Vec<Requirement>,
}

impl RequirementsSpec {
    pub fn new(
        title: impl Into<String>,
        overview: impl Into<String>,
        requirements: Vec<Requirement>,
    ) -> Self {
        Self {
            title: title.into(),
            overview: overview.into(),
            requirements,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Requirement {
    pub id: String,
    pub summary: String,
    pub acceptance_criteria: Vec<AcceptanceCriterion>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub file_hints: Vec<String>,
}

impl Requirement {
    pub fn new(
        id: impl Into<String>,
        summary: impl Into<String>,
        acceptance_criteria: Vec<AcceptanceCriterion>,
        file_hints: Vec<String>,
    ) -> Self {
        Self {
            id: id.into(),
            summary: summary.into(),
            acceptance_criteria,
            file_hints,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct AcceptanceCriterion {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub text: String,
}

impl AcceptanceCriterion {
    pub fn new(id: Option<String>, text: impl Into<String>) -> Self {
        Self {
            id,
            text: text.into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ProposedChanges {
    pub rationale: String,
    pub changes: Vec<ProposedChange>,
}

impl ProposedChanges {
    pub fn new(rationale: impl Into<String>, changes: Vec<ProposedChange>) -> Self {
        Self {
            rationale: rationale.into(),
            changes,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ProposedChange {
    pub requirement_id: String,
    pub summary: String,
    pub files: Vec<ChangeFile>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub notes: Vec<String>,
}

impl ProposedChange {
    pub fn new(
        requirement_id: impl Into<String>,
        summary: impl Into<String>,
        files: Vec<ChangeFile>,
        notes: Vec<String>,
    ) -> Self {
        Self {
            requirement_id: requirement_id.into(),
            summary: summary.into(),
            files,
            notes,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ChangeFile {
    pub path: String,
    pub change_type: ChangeType,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub rationale: String,
}

impl ChangeFile {
    pub fn new(
        path: impl Into<String>,
        change_type: ChangeType,
        rationale: impl Into<String>,
    ) -> Self {
        Self {
            path: path.into(),
            change_type,
            rationale: rationale.into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ChangeType {
    Create,
    Modify,
    Remove,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct TestPlan {
    pub summary: String,
    pub tasks: Vec<TestTask>,
}

impl TestPlan {
    pub fn new(summary: impl Into<String>, tasks: Vec<TestTask>) -> Self {
        Self {
            summary: summary.into(),
            tasks,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct TestTask {
    pub name: String,
    pub command: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub related_requirements: Vec<String>,
}

impl TestTask {
    pub fn new(
        name: impl Into<String>,
        command: impl Into<String>,
        related_requirements: Vec<String>,
    ) -> Self {
        Self {
            name: name.into(),
            command: command.into(),
            related_requirements,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct TestResults {
    pub plan: TestPlan,
    pub outcomes: Vec<TestOutcome>,
}

impl TestResults {
    pub fn new(plan: TestPlan, outcomes: Vec<TestOutcome>) -> Self {
        Self { plan, outcomes }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct TestOutcome {
    pub task_name: String,
    pub status: TestStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub details: Option<String>,
}

impl TestOutcome {
    pub fn new(task_name: impl Into<String>, status: TestStatus, details: Option<String>) -> Self {
        Self {
            task_name: task_name.into(),
            status,
            details,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TestStatus {
    Passed,
    Failed,
    Blocked,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ReviewFindings {
    pub summary: String,
    pub findings: Vec<ReviewFinding>,
}

impl ReviewFindings {
    pub fn new(summary: impl Into<String>, findings: Vec<ReviewFinding>) -> Self {
        Self {
            summary: summary.into(),
            findings,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ReviewFinding {
    pub severity: Severity,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub related_requirement: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub suggested_fix: Option<String>,
}

impl ReviewFinding {
    pub fn new(
        severity: Severity,
        message: impl Into<String>,
        related_requirement: Option<String>,
        suggested_fix: Option<String>,
    ) -> Self {
        Self {
            severity,
            message: message.into(),
            related_requirement,
            suggested_fix,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "snake_case")]
pub enum Severity {
    Info,
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct FormatterRun {
    pub command: String,
    pub status: FormatterStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output: Option<String>,
}

impl FormatterRun {
    pub fn skipped(command: impl Into<String>, reason: impl Into<String>) -> Self {
        Self {
            command: command.into(),
            status: FormatterStatus::Skipped,
            output: Some(reason.into()),
        }
    }

    pub fn success(command: impl Into<String>, output: Option<String>) -> Self {
        Self {
            command: command.into(),
            status: FormatterStatus::Succeeded,
            output,
        }
    }

    pub fn failed(command: impl Into<String>, output: impl Into<String>) -> Self {
        Self {
            command: command.into(),
            status: FormatterStatus::Failed,
            output: Some(output.into()),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FormatterStatus {
    Succeeded,
    Failed,
    Skipped,
}

pub fn derive_changes_from_spec(spec: &RequirementsSpec) -> ProposedChanges {
    let mut changes = Vec::new();
    for requirement in &spec.requirements {
        let files = derive_files_for_requirement(requirement);
        let notes = requirement
            .acceptance_criteria
            .iter()
            .map(|criterion| format!("Acceptance: {}", criterion.text))
            .collect();
        changes.push(ProposedChange::new(
            &requirement.id,
            &requirement.summary,
            files,
            notes,
        ));
    }
    ProposedChanges::new(
        format!("Derived from {} requirements", spec.requirements.len()),
        changes,
    )
}

fn derive_files_for_requirement(requirement: &Requirement) -> Vec<ChangeFile> {
    if requirement.file_hints.is_empty() {
        return vec![ChangeFile::new(
            "src/lib.rs",
            ChangeType::Modify,
            format!("Default target inferred for {}", requirement.id),
        )];
    }

    requirement
        .file_hints
        .iter()
        .map(|hint| {
            ChangeFile::new(
                hint,
                ChangeType::Modify,
                format!("Referenced by {}", requirement.id),
            )
        })
        .collect()
}

pub fn plan_tests_for_changes(changes: &ProposedChanges) -> TestPlan {
    let mut tasks = Vec::new();
    let mut related = Vec::new();
    for change in &changes.changes {
        related.push(change.requirement_id.clone());
    }
    if !changes.changes.is_empty() {
        tasks.push(TestTask::new("cargo test", "cargo test", related.clone()));
        tasks.push(TestTask::new("fmt check", "cargo fmt -- --check", related));
    }

    let summary = if tasks.is_empty() {
        "No tests required for documentation-only changes".to_string()
    } else {
        format!("Planned {} automated tasks", tasks.len())
    };
    TestPlan::new(summary, tasks)
}

pub fn merge_test_results(plan: &TestPlan, statuses: Vec<TestOutcome>) -> TestResults {
    let mut outcomes = Vec::new();
    for task in &plan.tasks {
        if let Some(existing) = statuses.iter().find(|status| status.task_name == task.name) {
            outcomes.push(existing.clone());
        } else {
            outcomes.push(TestOutcome::new(
                &task.name,
                TestStatus::Blocked,
                Some("No execution result available".to_string()),
            ));
        }
    }
    TestResults::new(plan.clone(), outcomes)
}

pub fn review_findings_from_results(
    changes: &ProposedChanges,
    results: &TestResults,
) -> ReviewFindings {
    let mut findings = Vec::new();

    for change in &changes.changes {
        if change
            .notes
            .iter()
            .any(|note| note.to_lowercase().contains("security"))
        {
            findings.push(ReviewFinding::new(
                Severity::High,
                format!(
                    "Security-related requirement {} needs dedicated review",
                    change.requirement_id
                ),
                Some(change.requirement_id.clone()),
                None,
            ));
        }
    }

    for outcome in &results.outcomes {
        if outcome.status == TestStatus::Failed {
            findings.push(ReviewFinding::new(
                Severity::Critical,
                format!("Test {} failed", outcome.task_name),
                None,
                outcome.details.clone(),
            ));
        }
        if outcome.status == TestStatus::Blocked {
            findings.push(ReviewFinding::new(
                Severity::Medium,
                format!("Test {} did not run", outcome.task_name),
                None,
                outcome.details.clone(),
            ));
        }
    }

    findings.sort_by(|a, b| b.severity.cmp(&a.severity));

    let summary = if findings.is_empty() {
        "No review findings".to_string()
    } else {
        format!("Identified {} findings", findings.len())
    };

    ReviewFindings::new(summary, findings)
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn derives_changes_with_default_file() {
        let spec = RequirementsSpec::new(
            "Spec",
            "Overview",
            vec![Requirement::new(
                "REQ-001",
                "Add feature",
                vec![
                    AcceptanceCriterion::new(None, "Works"),
                    AcceptanceCriterion::new(None, "Tested"),
                ],
                vec![],
            )],
        );

        let changes = derive_changes_from_spec(&spec);
        assert_eq!(changes.changes.len(), 1);
        let change = &changes.changes[0];
        assert_eq!(change.files.len(), 1);
        assert_eq!(change.files[0].path, "src/lib.rs");
    }

    #[test]
    fn plans_tests_for_changes() {
        let changes = ProposedChanges::new(
            "rationale",
            vec![ProposedChange::new(
                "REQ-001",
                "do it",
                vec![ChangeFile::new("src/main.rs", ChangeType::Modify, "reason")],
                vec![],
            )],
        );

        let plan = plan_tests_for_changes(&changes);
        assert_eq!(plan.tasks.len(), 2);
        assert!(plan.summary.contains("Planned"));
    }

    #[test]
    fn review_findings_cover_failed_tests() {
        let changes = ProposedChanges::new(
            "r",
            vec![ProposedChange::new(
                "REQ-001",
                "Work",
                vec![ChangeFile::new("src/lib.rs", ChangeType::Modify, "")],
                vec![],
            )],
        );
        let plan = plan_tests_for_changes(&changes);
        let results = merge_test_results(
            &plan,
            vec![TestOutcome::new(
                "cargo test",
                TestStatus::Failed,
                Some("compile error".to_string()),
            )],
        );
        let findings = review_findings_from_results(&changes, &results);
        assert!(findings.summary.contains("Identified"));
        assert!(!findings.findings.is_empty());
        assert!(matches!(findings.findings[0].severity, Severity::Critical));
    }
}
