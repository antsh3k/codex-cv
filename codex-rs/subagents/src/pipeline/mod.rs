//! Shared pipeline utilities for subagent data transformation and workflow orchestration.

pub mod types;
pub mod transform;
pub mod validation;

pub use types::*;
pub use transform::*;
pub use validation::*;