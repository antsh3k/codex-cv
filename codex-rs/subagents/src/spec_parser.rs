use crate::AgentSource;
use crate::DiagnosticLevel;
use crate::Subagent;
use crate::SubagentBuilder;
use crate::SubagentSpec;
use crate::TaskContext;
use crate::TaskContextError;
use crate::TypedSubagent;
use crate::pipeline::AcceptanceCriterion;
use crate::pipeline::Requirement;
use crate::pipeline::RequirementsSpec;
use anyhow::Result;
use anyhow::anyhow;
use once_cell::sync::Lazy;
use regex_lite::Regex;
use serde::Serialize;
use std::borrow::Cow;
use std::collections::HashSet;

static REQUIREMENT_LINE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^-\s*\[(?P<id>[A-Z0-9_-]+)\]\s*(?P<summary>.+)$").expect("valid regex")
});

static ACCEPTANCE_LINE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^(?P<id>AC-[0-9]{3,})?\s*:?\s*(?P<text>.+)$").expect("valid regex"));

static FILE_HINT_LINE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?i)^file[s]?\s*:\s*(?P<files>.+)$").expect("valid regex"));

const SPEC_PARSER_PROMPT: &str = r#"
You are the specification parser subagent. Read the user's request and produce a concise
requirements document. Ensure every requirement has an identifier (format `REQ-###`),
a summary, and at least one acceptance criterion. Capture any referenced files when
explicitly mentioned (e.g. `files: src/lib.rs`).

Return JSON following this schema:
- title: short title for the effort
- overview: one paragraph summary
- requirements: array of requirement objects with `id`, `summary`,
  `acceptance_criteria` (array of strings), and optional `file_hints` (array of strings).
"#;

#[derive(Default)]
pub struct SpecParserSubagent;

#[derive(Debug, Clone)]
pub struct SpecParserSeed {
    pub markdown: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct SpecParserOutput {
    pub spec: RequirementsSpec,
}

impl SpecParserSeed {
    pub fn new(markdown: impl Into<String>) -> Self {
        Self {
            markdown: markdown.into(),
        }
    }
}

impl SpecParserSubagent {
    pub fn subagent_spec() -> SubagentSpec {
        SubagentBuilder::new("spec-parser")
            .description(Some(
                "Parses natural language briefs into structured requirements".to_string(),
            ))
            .model(Some("gpt-5-codex".to_string()))
            .keywords(["requirements", "spec", "analysis"])
            .instructions(SPEC_PARSER_PROMPT)
            .source(AgentSource::Builtin)
            .build()
            .expect("valid spec parser definition")
    }

    fn load_seed(ctx: &TaskContext) -> Result<SpecParserSeed, anyhow::Error> {
        ctx.get_typed::<SpecParserSeed>()?
            .ok_or_else(|| anyhow!("SpecParserSeed not present in task context"))
    }
}

impl Subagent for SpecParserSubagent {
    fn spec(&self) -> Cow<'_, SubagentSpec> {
        Cow::Owned(Self::subagent_spec())
    }
}

impl TypedSubagent for SpecParserSubagent {
    type Input = SpecParserSeed;
    type Output = SpecParserOutput;

    fn prepare(&self, ctx: &TaskContext) -> Result<Self::Input> {
        Self::load_seed(ctx)
    }

    fn execute(&self, ctx: &mut TaskContext, input: Self::Input) -> Result<Self::Output> {
        let parsed = parse_spec_markdown(&input.markdown).map_err(|err| {
            let _ = ctx.push_diagnostic(
                DiagnosticLevel::Error,
                format!("Spec parsing failed: {err}"),
            );
            err
        })?;
        ctx.push_diagnostic(
            DiagnosticLevel::Info,
            format!("Parsed {} requirements", parsed.requirements.len()),
        )?;
        Ok(SpecParserOutput { spec: parsed })
    }

    fn finalize(&self, ctx: &mut TaskContext, output: Self::Output) -> Result<()> {
        ctx.insert_typed(output.spec.clone())?;
        ctx.set_scratchpad(
            "subagents.spec_parser.output",
            serde_json::to_value(&output).map_err(TaskContextError::Serialization)?,
        )?;
        Ok(())
    }
}

fn parse_spec_markdown(markdown: &str) -> Result<RequirementsSpec> {
    let mut title = String::new();
    let mut overview_lines = Vec::new();
    let mut requirements = Vec::new();
    let mut in_requirements_section = false;
    let mut current: Option<RequirementBuilder> = None;
    let mut seen_ids = HashSet::new();

    for raw_line in markdown.lines() {
        let line = raw_line.trim_end();
        if line.starts_with('#') {
            if line.starts_with("##") {
                let heading = line.trim_start_matches('#').trim();
                if heading.eq_ignore_ascii_case("requirements") {
                    if let Some(builder) = current.take() {
                        requirements.push(builder.finish()?);
                    }
                    in_requirements_section = true;
                    continue;
                }
            } else if title.is_empty() {
                title = line.trim_start_matches('#').trim().to_string();
                continue;
            }
        }

        if !in_requirements_section {
            if !line.trim().is_empty() {
                overview_lines.push(line.trim().to_string());
            }
            continue;
        }

        let trimmed = line.trim();

        if let Some(captures) = REQUIREMENT_LINE.captures(trimmed) {
            if let Some(builder) = current.take() {
                requirements.push(builder.finish()?);
            }
            let id = captures
                .name("id")
                .expect("id capture")
                .as_str()
                .to_string();
            validate_requirement_id(&id)?;
            if !seen_ids.insert(id.clone()) {
                return Err(anyhow!("Duplicate requirement id {id}"));
            }
            let summary = captures
                .name("summary")
                .expect("summary capture")
                .as_str()
                .trim()
                .to_string();
            current = Some(RequirementBuilder::new(id, summary));
            continue;
        }

        if let Some(builder) = current.as_mut() {
            if !trimmed.starts_with('-') {
                continue;
            }
            let inner = trimmed.trim_start_matches('-').trim();

            if let Some(file_caps) = FILE_HINT_LINE.captures(inner) {
                let files = file_caps.name("files").expect("files").as_str();
                let hints = files
                    .split(',')
                    .map(|f| f.trim().to_string())
                    .filter(|f| !f.is_empty());
                for hint in hints {
                    builder.file_hints.push(hint);
                }
                continue;
            }

            if let Some(ac_caps) = ACCEPTANCE_LINE.captures(inner) {
                let id = ac_caps.name("id").map(|m| m.as_str().to_string());
                let text = ac_caps
                    .name("text")
                    .map(|m| m.as_str().trim().to_string())
                    .unwrap_or_default();
                if text.is_empty() {
                    return Err(anyhow!("Acceptance criterion missing text"));
                }
                builder
                    .acceptance_criteria
                    .push(AcceptanceCriterion::new(id, text));
                continue;
            }

            if !inner.is_empty() {
                builder
                    .acceptance_criteria
                    .push(AcceptanceCriterion::new(None, inner.to_string()));
            }
        }
    }

    if let Some(builder) = current.take() {
        requirements.push(builder.finish()?);
    }

    if requirements.is_empty() {
        return Err(anyhow!("No requirements found"));
    }

    let overview = if overview_lines.is_empty() {
        "No overview provided".to_string()
    } else {
        overview_lines.join(" ")
    };

    Ok(RequirementsSpec::new(
        if title.is_empty() { "Untitled" } else { &title },
        overview,
        requirements,
    ))
}

fn validate_requirement_id(id: &str) -> Result<()> {
    if !id.starts_with("REQ-") {
        return Err(anyhow!("Requirement id {id} must start with REQ-"));
    }
    let suffix = &id[4..];
    if suffix.is_empty() || suffix.chars().any(|c| !c.is_ascii_digit()) {
        return Err(anyhow!("Requirement id {id} must end in digits"));
    }
    Ok(())
}

struct RequirementBuilder {
    id: String,
    summary: String,
    acceptance_criteria: Vec<AcceptanceCriterion>,
    file_hints: Vec<String>,
}

impl RequirementBuilder {
    fn new(id: String, summary: String) -> Self {
        Self {
            id,
            summary,
            acceptance_criteria: Vec::new(),
            file_hints: Vec::new(),
        }
    }

    fn finish(self) -> Result<Requirement> {
        if self.acceptance_criteria.is_empty() {
            return Err(anyhow!(
                "Requirement {} missing acceptance criteria",
                self.id
            ));
        }
        Ok(Requirement::new(
            self.id,
            self.summary,
            self.acceptance_criteria,
            self.file_hints,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::TaskContext;

    const SAMPLE: &str = r#"
# Feature Rollout

The product team needs an authenticated export feature.

## Requirements
- [REQ-001] Add export command
  - AC-001: CLI exposes `codex export`
  - AC-002: command requires auth token
  - files: cli/src/commands/export.rs, core/src/export.rs
- [REQ-002] Log export telemetry
  - AC-003: record success and failure events
  - file: core/src/telemetry.rs
"#;

    #[test]
    fn parses_markdown_into_structured_spec() {
        let mut ctx = TaskContext::new();
        ctx.insert_typed(SpecParserSeed::new(SAMPLE.to_string()))
            .unwrap();
        let agent = SpecParserSubagent;
        let input = agent.prepare(&ctx).unwrap();
        let output = agent.execute(&mut ctx, input).unwrap();
        agent.finalize(&mut ctx, output.clone()).unwrap();

        let snapshot = serde_json::to_string_pretty(&output).unwrap();
        insta::assert_snapshot!("spec_parser_basic", snapshot);
    }

    #[test]
    fn rejects_missing_acceptance_criteria() {
        let markdown = "## Requirements\n- [REQ-123] Missing criteria";
        let err = parse_spec_markdown(markdown).unwrap_err();
        assert!(err.to_string().contains("missing acceptance"));
    }
}
