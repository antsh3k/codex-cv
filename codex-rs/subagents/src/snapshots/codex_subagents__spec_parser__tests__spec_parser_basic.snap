---
source: subagents/src/spec_parser.rs
expression: snapshot
---
{
  "spec": {
    "title": "Feature Rollout",
    "overview": "The product team needs an authenticated export feature.",
    "requirements": [
      {
        "id": "REQ-001",
        "summary": "Add export command",
        "acceptance_criteria": [
          {
            "id": "AC-001",
            "text": "CLI exposes `codex export`"
          },
          {
            "id": "AC-002",
            "text": "command requires auth token"
          }
        ],
        "file_hints": [
          "cli/src/commands/export.rs",
          "core/src/export.rs"
        ]
      },
      {
        "id": "REQ-002",
        "summary": "Log export telemetry",
        "acceptance_criteria": [
          {
            "id": "AC-003",
            "text": "record success and failure events"
          }
        ],
        "file_hints": [
          "core/src/telemetry.rs"
        ]
      }
    ]
  }
}
